pkg/suggestion/NASenvelopenet/generate.py

```diff
@@ -0,0 +1,487 @@
+        #print("Input channels: " + str(inputchannels))
+        width = math.pow(2, stagenum) * inputchannels
+        #print("W : " + str(width))
+        if self.parameter_limits[stagenum]:
+            """ Parameter limiting: Calculate output of the internal filters such that
+            overall  params is maintained constant
+            """
+            layers = float(len(stage))
+
+            outputs = int((width / (layers - 2.0)) *
+                          (math.pow(layers - 1.0, 0.5) - 1))
+            
+            #print("outputs : " + str(outputs))
+        #print(stage)
+        lidx = 0
+        for layer in stage:
+            #print(lidx)
+            #print(len(stage) - 1)
+            if "widener" in layer:
+                #print("Widener")
+                lidx += 1
+                continue
+            if lidx == len(stage) - \
+                    1 or self.parameter_limits[stagenum] is False:
+                #print("Setting outputs to W")
+                layer["outputs"] = int(width)
+            elif "filters" in layer:
+                #print("Limiting outputs")
+                layer["outputs"] = outputs
+            lidx += 1
+
+    def select_prunable(self, stagecellnames, samples, worst_case=False):
+        print(stagecellnames)
+        #samples = self.filter_samples(samples)
+        measurements = {}
+        for sample in samples:
+            if sample == '':
+                continue
+            sample = self.remove_logging(sample)
+            filt, value = self.get_filter_sample(sample)
+
+            # Prune only filters in this stage
+            if filt not in stagecellnames:
+                continue
+
+            if filt not in measurements:
+                measurements[filt] = []
+            measurements[filt].append(value)
+
+        #print("Stage cell names" + str(stagecellnames))
+        #print("Filter in samples " + str(list(measurements.keys())))
+        # Rank variances, select filters to prune
+        # Use last variance reading
+        variances = {}
+        for filt in measurements:
+            variances[filt] = measurements[filt][-1]
+
+        if worst_case:
+            print("WARNING: -------GENERATING WORST CASE NET---------")
+            reverse = True
+        else:
+            reverse = False
+        svariances = sorted(variances, key=variances.get, reverse=reverse)
+        # Count number of cells in each layer
+        #print("All variances: " + str(variances))
+        #print("Sorted variances: " + str(svariances))
+        cellcount = {}
+        for cellbr in variances:
+            cellidx = cellbr.split("/")[0].lstrip("Cell")
+            if cellidx not in cellcount:
+                cellcount[cellidx] = 0
+            cellcount[cellidx] += 1
+
+        #print(cellcount)
+        # Make sure we do not prune all cells in one layer
+        prunedcount = {}
+        prune = []
+        for svariance in svariances:
```

Comment:
Variance metric is dropped from the final implementation of Envelopenet, you may use only the mean(l2-norm based) metric.